### PR TITLE
3d: Batch processing and Linux binary package

### PR DESCRIPTION
**Pull request by tahina-pro**
New features:
* I added a batch postprocessing phase to the 3d tool itself: with `--batch`, the produced F* files are now verified and C code is generated. Basically, the postprocessing behavior of `everparse.bat`, and more, now moved to `3d --batch`
* Several `.3d` files can be given as arguments to `3d`.  If at least two files are provided, then `--module_name` is forbidden and the module name of each `.3d` file is its basename. Then, with `--batch`, a common shared `EverParse.h` is generated containing the union of all helpers needed by each `.3d` file.
* With `--batch`, if `Foo.3d` is accompanied with `Foo.3d.copyright.txt` containing well-formed C comments, then those comments are prepended to the `Foo`-specific C files and headers. Comments for `EverParse.h` are handled separately through a fixed `src/3d/copyright.txt`
* I added an option, `--clang_format`, to process all .c and .h files through clang-format, with an option `--clang_format_executable` to specify the location of clang-format.
* I added a [clang-format binary snapshot](http://llvm.org/builds/) to the Windows binary package, so that `--clang_format` is now enabled by default with `everparse.bat`
* I built a Linux binary package, with a Bash script `everparse.sh` as a counterpart to Windows' `.bat`,  enabling `--clang_format` by default if it is reachable from the `PATH` (no clang-format binary snapshot available on Linux)
* I documented all those features, you can `make -C doc html` and read

Changes to existing features:
* The arguments syntax for `everparse.bat` is now that of `3d`. Files now have to be provided with their full name and extension. No need to be in the same directory as the `.3d` file; in particular, `everparse.bat` can now be called with `.3d` files located in different directories. Windows-style options such as `/F` and peculiarities such as module-without-filename-extension disappear.
* `--odir` now defaults to `.` instead of `out`
* (Unrelated) I had to disable some tests which are no longer valid because of double fetches: enum-tagged unions. I detected those failing tests through the fact that I added a `--batch` test to the CI.

### CODE REVIEW

File: src/3d/Binding.fst [hunk 1]
```diff
@@ -110,7 +103,19 @@ let nullary_macro t = { macro_arguments_t = []; macro_result_t = t }
 /// global_env maps top-level identifiers to their corresponding declaration
 ///  -- maps type identifiers to decl_attributes
 ///  -- maps macro names to their types
-let global_env = H.t ident' (decl & either decl_attributes macro_signature)
+noeq
+type global_env = {
+  ge_h: H.t ident' (decl & either decl_attributes macro_signature);
+  ge_fd: field_num_ops_t;
```
Comment (nikswamy) [inline src/3d/Binding.fst:109]: Looks like the field num table is now carried as part of the environment. Sounds good, in general. Can you say a bit about what motivated this change? Would also be good to update the comment on the global_env type just above.

File: src/3d/Options.fst [hunk 2]
```diff
@@ -9,33 +9,80 @@ let input_file : ref (list string) = alloc []
...
 
-let options =
+(* We would like to parse --help as an option, but this would
+   require to recurse on the definition of the list of options. To
+   avoid that, we define this list as mutable, and then we add --help
+   after the fact. *)
+
+let options0 =
   let open FStar.Getopt in
   [(noshort, "module_name", OneArg ((fun mname -> module_name := Some mname), "module name"), "module name to use for the output file");
```
Comment (nikswamy) [inline src/3d/Options.fst:24]: Wonder if we should just ditch the module_name option and be uniform about using the file name prefix as the module name. We could do that in a separate PR.

File: src/3d/Binding.fst (hunk 1 above)
Comment (tahina-pro) [reply src/3d/Binding.fst:109]: This class of changes is to make sure that the field table and other global state is not shared among different .3d files provided on the command line.

File: src/3d/Options.fst (hunk 2 above)
Comment (tahina-pro) [reply src/3d/Options.fst:24]: Yes, we could. What was the initial intent behind `module_name`?

Comment (nikswamy) [reply src/3d/Options.fst:24]: No good reason, I think. Not sure why I didn't go for the file name thing to begin with.

File: doc/3d.rst [hunk 3]
```diff
@@ -54,12 +58,233 @@ The 3d Dependent Data Description language
...
+  ``.c`` and ``.h`` files. This is toggled by default.
+
+* ``--no_batch``
+
+  Do not verify the produced F* ``.fst`` and ``.fsti`` files or
+  generate C ``.c`` and ``.h`` files.
+
+* ``--clang_format``
+
+  Call ``clang-format`` on extracted .c/.h files. This option
```
[the remark this answers is not in the document]
Comment (tahina-pro) [reply doc/3d.rst:119]: this is to shorten the command line when invoking `3d`. Indeed `--clang_format` only makes sense with `--batch` on because `.c`/`.h` files are produced only with `--batch` (otherwise only `.fst*` files are produced)

File: src/package/README.Windows.pkg [hunk 4]
```diff
@@ -0,0 +1,51 @@
+EverParse: verified parsing for binary data formats
```
[the remark this answers is not in the document]
Comment (tahina-pro) [reply src/package/README.Windows.pkg:1]: `src/package/README.*.pkg` are for the binary package, whereas `README.md` is for the GitHub clone, for those who want to build from source. The problem is that invocation of EverParse is different in the two cases, because:
* the `everparse.bat` or `everparse.sh` designed for the package will not work from the GitHub clone since it assumes locations for F* and KReMLin that are valid only for the package (and no, those scripts are not designed for users to supersede `FSTAR_HOME` etc.)
* users of the binary package could run `3d` directly from the package, but they would need to manually set up their environment with `FSTAR_HOME` etc., which would defeat the purpose of a standalone binary package

File: src/3d/ocaml/Batch.ml [hunk 5]
```diff
@@ -0,0 +1,402 @@
...
+      Printf.sprintf "%s.c" modul;
+      Printf.sprintf "%s.h" modul;
+      Printf.sprintf "%sWrapper.c" modul;
+      Printf.sprintf "%sWrapper.h" modul;
+    ]
+  end
+  
+(* Collect all produced .c and .h files *)
+
+let collect_file
```
[the remark this answers is not in the document]
Comment (tahina-pro) [reply src/3d/ocaml/Batch.ml:324]: If the `.3d` file is empty or has no `entrypoint`, then KReMLin will produce no `.c` file. Should `--batch` fail in such cases?

File: src/3d/.gitattributes [hunk 6]
```diff
@@ -0,0 +1,3 @@
+# Bash scripts MUST have Unix line endings
```
[the remark this answers is not in the document]
Comment (tahina-pro) [reply src/3d/.gitattributes:1]: Yes, in fact none of them are strictly necessary. I was trying `bash src/3d/version.sh`, which failed on Windows, but making it executable and running it directly without an explicit `bash` invocation finally made it work.

File: src/3d/Options.fst [hunk 7]
```diff
@@ -9,33 +9,80 @@ let input_file : ref (list string) = alloc []
...
+  match String.split ['/'] fn2 with
+  | [] -> fn2
+  | l2 -> list_last l2
+
 let split_3d_file_name fn =
-  match String.split ['.'] fn with
+  match String.split ['.'] (basename fn) with
   | [name;extension] ->
     // FStar.IO.print_string
     //   (Printf.sprintf "filename = %s; name=%s; extension=%s"
```
[the remark this answers is not in the document]
Comment (tahina-pro) [reply src/3d/Options.fst:88]: it is intended to fail anyway because the basename of the file, which will be reused as a component of a C identifier, should not have `.`
Maybe we should have an explicit validation step on what we expect to be a valid module name.

File: src/3d/ocaml/Batch.ml [hunk 8]
```diff
@@ -0,0 +1,402 @@
...
+
+let filename_concat = Filename.concat
+
+let run_cmd_gen reconcat prog args =
+  let cmd = String.concat " " (prog :: args) in
+  print_endline (Printf.sprintf "Running: %s" cmd);
+  let args' = Array.of_list (if reconcat then prog :: args else args) in (* FIXME: WHY WHY WHY do I need to recons the prog in front of the args? *)
+  let pid = Unix.create_process prog args' Unix.stdin Unix.stdout Unix.stderr in
+  let (_, res) = Unix.waitpid [] pid in
+  let res = code_of_exit res in
```
[the remark this answers is not in the document]
Comment (tahina-pro) [reply src/3d/ocaml/Batch.ml:15]: Thanks a lot for your suggestion. In fact, with Process, I claim that I should use `execute` (because I don't need to capture output and so I may not want to have output retained in memory), but it is not exposed to the `.mli`.

File: src/3d/.gitattributes (hunk 6 above)
[the remark this answers is not in the document]
Comment (msprotz) [reply src/3d/.gitattributes:1]: a good idea nonetheless to have a .gitattributes given the wide prevalence of mis-configurations on Windows

File: src/3d/Options.fst (hunk 7 above)
[the remark this answers is not in the document]
Comment (msprotz) [reply src/3d/Options.fst:88]: early validation sounds good rather than late-failure

File: src/3d/ocaml/Batch.ml (hunk 5 above)
[the remark this answers is not in the document]
Comment (msprotz) [reply src/3d/ocaml/Batch.ml:324]: sounds good to fail with a descriptive error message, e.g. "no .c file generated by kremlin, you most likely forgot to mark an `entrypoint`"

**Comment by tahina-pro** [comment]
Thank you both for your feedback!
At this point I addressed part of @msprotz 's OS-related comments (filename, process) by moving some (but not all) primitives to a new OCaml module, `OS.ml`. I could go on and rewrite `Batch.ml` in F* with an appropriately enriched `OS.fsti`. At the very best, I could even upgrade `OS.*` to an appropriate `FStar.OS` interface and OCaml implementation to be included into the F* standard library. What do you think?

File: src/3d/ocaml/Batch.ml [hunk 9]
```diff
@@ -0,0 +1,402 @@
...
+         let ln = Str.global_replace regexp_FILENAME filename ln in
+         output_line channel_out ln;
+         aux ()
+  in
+  aux ();
+  close_in cin
+
+(* Copyright headers *)
+
+let cat
```
[the remark this answers is not in the document]
Comment (tahina-pro) [reply src/3d/ocaml/Batch.ml]: (Moved to `src/3d/OS.ml`, function `cat`)
I could have used `BatIO.copy`, but my implementation also implicitly changes line endings to those of the running OS, which is what we want.

File: src/3d/ocaml/Batch.ml (hunk 5 above)
[the remark this answers is not in the document]
Comment (tahina-pro) [reply src/3d/ocaml/Batch.ml:251]: Solved by #25

File: src/3d/ocaml/Batch.ml [hunk 10]
```diff
@@ -0,0 +1,402 @@
...
+(* paths *)
+let fstar_home = Sys.getenv "FSTAR_HOME"
+let kremlin_home = Sys.getenv "KREMLIN_HOME"
+let kremlib = filename_concat kremlin_home "kremlib"
+let qd_home = Sys.getenv "QD_HOME"
+let lowparse_home = filename_concat (filename_concat qd_home "src") "lowparse"
+let ddd_home = filename_concat (filename_concat qd_home "src") "3d"
+let ddd_prelude_home = filename_concat (filename_concat (filename_concat qd_home "src") "3d") "prelude"
+
+let regexp_fslash = Str.regexp "/"
```
[the remark this answers is not in the document]
Comment (tahina-pro) [reply src/3d/ocaml/Batch.ml]: Tracked by #19 , just fixed by 140adb100e7e369c2441f1dba86bbf5a7e379860
